File: app/src/main/java/com/hover/stax/actions/ActionSelect.java
package com.hover.stax.actions;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.drawable.Drawable;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.AutoCompleteTextView;
import android.widget.LinearLayout;
import android.widget.RadioButton;
import android.widget.RadioGroup;
import android.widget.TextView;

import androidx.core.graphics.drawable.RoundedBitmapDrawable;
import androidx.core.graphics.drawable.RoundedBitmapDrawableFactory;

import com.hover.sdk.actions.HoverAction;
import com.hover.stax.R;
import com.hover.stax.databinding.ActionSelectBinding;
import com.hover.stax.utils.UIHelper;
import com.hover.stax.views.AbstractStatefulInput;
import com.hover.stax.views.StaxDropdownLayout;
import com.squareup.picasso.Picasso;
import com.squareup.picasso.Target;

import java.util.ArrayList;
import java.util.List;

import timber.log.Timber;

import static com.hover.stax.utils.Constants.size55;

public class ActionSelect extends LinearLayout implements RadioGroup.OnCheckedChangeListener, Target {
    private static String TAG = "ActionSelect";

    private StaxDropdownLayout dropdownLayout;
    private AutoCompleteTextView dropdownView;
    private TextView radioHeader;
    private RadioGroup isSelfRadio;

    private List<HoverAction> actions;
    private HoverAction highlightedAction;
    private HighlightListener highlightListener;

    private ActionSelectBinding binding;

    public ActionSelect(Context context, AttributeSet attrs) {
        super(context, attrs);

        binding = ActionSelectBinding.inflate(LayoutInflater.from(context), this, true);

        init();
    }

    private void init() {
        dropdownLayout = binding.actionDropdownInput;
        dropdownView = binding.actionDropdownInput.autoCompleteTextView;
        radioHeader = binding.header;
        isSelfRadio = binding.isSelfRadioGroup;
        this.setVisibility(GONE);
    }

    public void updateActions(List<HoverAction> filteredActions) {
        Timber.e("Updating to %s", filteredActions);
        this.setVisibility(filteredActions == null || filteredActions.size() <= 0 ? View.GONE : View.VISIBLE);
        if (filteredActions == null || filteredActions.size() <= 0) return;

        actions = filteredActions;
        highlightedAction = null;
        List<HoverAction> uniqRecipientActions = sort(filteredActions);

        ActionDropdownAdapter actionDropdownAdapter = new ActionDropdownAdapter(uniqRecipientActions, getContext());
        dropdownView.setAdapter(actionDropdownAdapter);
        dropdownView.setOnItemClickListener((adapterView, view2, pos, id) -> selectRecipientNetwork((HoverAction) adapterView.getItemAtPosition(pos)));
        Timber.e("uniq recipient networks %s", uniqRecipientActions.size());
        dropdownLayout.setVisibility(showRecipientNetwork(uniqRecipientActions) ? VISIBLE : GONE);
        Timber.e(actions.get(0).transaction_type);

        radioHeader.setText(actions.get(0).transaction_type.equals(HoverAction.AIRTIME) ? R.string.airtime_who_header : R.string.send_who_header);
    }

    public static List<HoverAction> sort(List<HoverAction> actions) {
        ArrayList<Integer> uniqRecipInstIds = new ArrayList<>();
        ArrayList<HoverAction> uniqRecipActions = new ArrayList<>();
        for (HoverAction a : actions) {
            if (!uniqRecipInstIds.contains(a.recipientInstitutionId())) {
                uniqRecipInstIds.add(a.recipientInstitutionId());
                uniqRecipActions.add(a);
            }
        }
        return uniqRecipActions;
    }

    private boolean showRecipientNetwork(List<HoverAction> actions) {
        return actions.size() > 1 || (actions.size() == 1 && !actions.get(0).isOnNetwork());
    }

    public void selectRecipientNetwork(HoverAction action) {
        if (action.equals(highlightedAction)) return;

        setDropDownValue(action);
        setRadioValuesIfRequired(action);
    }

    private void setRadioValuesIfRequired(HoverAction action) {
        dropdownLayout.setState(null, AbstractStatefulInput.SUCCESS);
        List<HoverAction> options = getWhoMeOptions(action.recipientInstitutionId());
        if (options.size() == 1) {
            if (!options.get(0).requiresRecipient())
                dropdownLayout.setState(getContext().getString(R.string.self_only_money_warning), AbstractStatefulInput.INFO);
            selectAction(action);
            isSelfRadio.setVisibility(GONE);
            radioHeader.setVisibility(GONE);
        } else
            createRadios(options);
    }

    private void setDropDownValue(HoverAction a) {
        dropdownView.setText(a.toString(), false);
        UIHelper.loadPicasso(getContext().getString(R.string.root_url) + a.to_institution_logo, size55, this);
    }

    public void selectAction(HoverAction a) {
        Timber.e("selecting action %s", a);
        highlightedAction = a;
        if (highlightListener != null) highlightListener.highlightAction(a);
    }

    public void setListener(HighlightListener hl) {
        highlightListener = hl;
    }

    public void setState(String message, int state) {
        dropdownLayout.setState(message, state);
    }

    private List<HoverAction> getWhoMeOptions(int recipientInstId) {
        List<HoverAction> options = new ArrayList<>();
        if (actions == null) return options;
        for (HoverAction a : actions) {
            if (a.recipientInstitutionId() == recipientInstId && !options.contains(a))
                options.add(a);
        }
        return options;
    }

    private void createRadios(List<HoverAction> actions) {
        isSelfRadio.removeAllViews();
        isSelfRadio.clearCheck();

        if(!actions.isEmpty()) {
            for (int i = 0; i < actions.size(); i++) {
                HoverAction a = actions.get(i);
                RadioButton radioButton = (RadioButton) LayoutInflater.from(getContext()).inflate(R.layout.stax_radio_button, null);
                radioButton.setText(a.getPronoun(getContext()));
                radioButton.setId(i);
                isSelfRadio.addView(radioButton);
            }
        }

        isSelfRadio.setOnCheckedChangeListener(this);
        isSelfRadio.check(highlightedAction != null ? actions.indexOf(highlightedAction) : 0);
        isSelfRadio.setVisibility(actions.size() > 1 ? VISIBLE : GONE);
        radioHeader.setVisibility(actions.size() > 1 ? VISIBLE : GONE);
    }

    @Override
    public void onCheckedChanged(RadioGroup group, int checkedId) {
        if (checkedId == -1 && actions != null) return;
        selectAction(actions.get(checkedId));
    }

    @Override
    public void onBitmapLoaded(Bitmap bitmap, Picasso.LoadedFrom from) {
        RoundedBitmapDrawable d = RoundedBitmapDrawableFactory.create(getContext().getResources(), bitmap);
        d.setCircular(true);
        dropdownView.setCompoundDrawablesRelativeWithIntrinsicBounds(d, null, null, null);
    }

    @Override
    public void onBitmapFailed(Exception e, Drawable errorDrawable) {
        Timber.e(e.getLocalizedMessage());
    }

    @Override
    public void onPrepareLoad(Drawable placeHolderDrawable) {
        dropdownView.setCompoundDrawablesRelativeWithIntrinsicBounds(R.drawable.ic_grey_circle_small, 0, 0, 0);
    }

    public interface HighlightListener {
        void highlightAction(HoverAction a);
    }
}

File: app/src/main/java/com/hover/stax/actions/ActionSelect.kt
package com.hover.stax.actions

import android.content.Context
import android.graphics.drawable.Drawable
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.widget.LinearLayout
import android.widget.RadioButton
import android.widget.RadioGroup
import com.bumptech.glide.request.target.CustomTarget
import com.bumptech.glide.request.transition.Transition
import com.hover.sdk.actions.HoverAction
import com.hover.stax.R
import com.hover.stax.databinding.ActionSelectBinding
import com.hover.stax.utils.UIHelper
import com.hover.stax.views.AbstractStatefulInput
import timber.log.Timber


class ActionSelect(context: Context, attrs: AttributeSet) : LinearLayout(context, attrs), RadioGroup.OnCheckedChangeListener {
    private var allActions: List<HoverAction>? = null
    private var uniqueInstitutions: List<HoverAction>? = null
    private var highlightedAction: HoverAction? = null
    private var highlightListener: HighlightListener? = null

    private var _binding: ActionSelectBinding? = null
    private val binding get() = _binding!!

    init {
        _binding = ActionSelectBinding.inflate(LayoutInflater.from(context), this, true)
        createListeners()
        visibility = GONE
    }

    private fun createListeners() {
        binding.actionDropdown.autoCompleteTextView.setOnItemClickListener { parent, _, position, _ ->
            selectRecipientNetwork(parent.getItemAtPosition(position) as HoverAction)
        }
    }

    fun updateActions(filteredActions: List<HoverAction>) {
        visibility = if (filteredActions.isEmpty()) GONE else VISIBLE
        if (filteredActions.isEmpty()) return

        allActions = filteredActions
        if (!filteredActions.contains(highlightedAction))
            highlightedAction = null

        uniqueInstitutions = sort(filteredActions)
        val actionDropdownAdapter = ActionDropdownAdapter(uniqueInstitutions!!, context)

        binding.header.setText(if (uniqueInstitutions!!.firstOrNull()?.transaction_type == HoverAction.AIRTIME) R.string.airtime_who_header else R.string.send_who_header)
        binding.actionDropdown.visibility = if (showRecipientNetwork(uniqueInstitutions!!)) View.VISIBLE else View.GONE
        binding.actionDropdown.autoCompleteTextView.setAdapter(actionDropdownAdapter)
    }

    fun sort(actions: List<HoverAction>): List<HoverAction> = actions.distinctBy { it.to_institution_id }.toList()

    private fun showRecipientNetwork(actions: List<HoverAction>): Boolean {
        return actions.size > 1 || (actions.size == 1 && !actions.first().isOnNetwork)
    }

    fun selectRecipientNetwork(action: HoverAction) {
        if (action == highlightedAction) return

        setState(null, AbstractStatefulInput.SUCCESS)
        binding.actionDropdown.autoCompleteTextView.setText(action.toString(), false)

        val target = object : CustomTarget<Drawable>() {
            override fun onResourceReady(resource: Drawable, transition: Transition<in Drawable>?) {
                binding.actionDropdown.autoCompleteTextView.setCompoundDrawablesRelativeWithIntrinsicBounds(resource, null, null, null)
                binding.actionDropdown.autoCompleteTextView.invalidate()
            }

            override fun onLoadCleared(placeholder: Drawable?) {
                binding.actionDropdown.autoCompleteTextView.setCompoundDrawablesRelativeWithIntrinsicBounds(R.drawable.ic_grey_circle_small, 0, 0, 0)
            }
        }

        UIHelper.loadImage(context, context.getString(R.string.root_url).plus(action.to_institution_logo), target)

        val options = getWhoMeOptions(action.to_institution_id)
        if (options.size == 1) selectOnlyOption(options.first())
        else createOptions(options)
    }

    private fun getWhoMeOptions(recipientInstId: Int): List<HoverAction> {
        return allActions?.filter { it.to_institution_id == recipientInstId } ?: emptyList()
    }

    private fun selectOnlyOption(option: HoverAction) {
        binding.header.visibility = GONE
        binding.isSelfRadio.visibility = GONE
        if (!option.requiresRecipient())
            setState(context.getString(R.string.self_only_money_warning), AbstractStatefulInput.INFO)
        selectAction(option)
    }

    private fun createOptions(recipientInstitutionActions: List<HoverAction>) {
        val radioVisibility = if (recipientInstitutionActions.size > 1) VISIBLE else GONE
        binding.header.visibility = radioVisibility
        binding.isSelfRadio.removeAllViews()
        binding.isSelfRadio.clearCheck()
        binding.isSelfRadio.visibility = radioVisibility
        binding.isSelfRadio.setOnCheckedChangeListener(this@ActionSelect)

        recipientInstitutionActions.forEachIndexed { index, action -> addOption(index, action) }
    }

    private fun addOption(index: Int, action: HoverAction) {
        val radioButton = (LayoutInflater.from(context).inflate(R.layout.stax_radio_button, null) as RadioButton).apply {
            text = action.getPronoun(context)
            id = action.id
            isChecked = index == 0
        }
        binding.isSelfRadio.addView(radioButton)
    }

    private fun selectAction(action: HoverAction) {
        highlightedAction = action
        highlightListener?.highlightAction(action)
    }

    fun setState(message: String?, state: Int) = binding.actionDropdown.setState(message, state)

    fun setListener(listener: HighlightListener) {
        highlightListener = listener
    }

    override fun onCheckedChanged(group: RadioGroup?, checkedId: Int) {
        if (checkedId == -1 || allActions.isNullOrEmpty() || allActions?.find { it.id == checkedId } == null) return
        val a = allActions!!.find { it.id == checkedId }!!
        selectAction(a)
    }

    interface HighlightListener {
        fun highlightAction(action: HoverAction?)
    }

}